src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs

```diff
@@ -20,99 +19,319 @@ public abstract class FileSystemWatcherTest : FileCleanupTestBase
+                    };
+                    // Most intermittent failures in FSW are caused by either a shortage of resources (e.g. inotify instances)
+                    // or by insufficient time to execute (e.g. CI gets bogged down). Immediately re-running a failed test
+                    // won't resolve the first issue, so we wait a little while hoping that things clear up for the next run.
+                    Thread.Sleep(500); 
+                }
 
-                watcher.Path = Path.GetFullPath(dir.Path);
-                watcher.Filter = "*";
-                watcher.NotifyFilter = changeFilers;
-                watcher.IncludeSubdirectories = true;
-                watcher.EnableRaisingEvents = true;
+                result = ExecuteAndVerifyEvents(watcher, expectedEvents, action, attemptsCompleted == attempts, expectedPaths);
 
-                using (var firstDir = new TempDirectory(Path.Combine(dir.Path, "dir1")))
-                {
-                    ExpectEvent(createdOccurred, "dir1 created");
+                if (cleanup != null)
+                    cleanup();
+            }
+        }
+
+        /// <summary>
+        /// Helper for the ExpectEvent function. 
+        /// </summary>
+        /// <param name="watcher">The FileSystemWatcher to test</param>
+        /// <param name="expectedEvents">All of the events that are expected to be raised by this action</param>
+        /// <param name="action">The Action that will trigger events.</param>
+        /// <param name="assertExpected">True if results should be asserted. Used if there is no retry.</param>
+        /// <param name="expectedPath"> Adds path verification to all expected events.</param>
+        /// <returns>True if the events raised correctly; else, false.</returns>
+        private static bool ExecuteAndVerifyEvents(FileSystemWatcher watcher, WatcherChangeTypes expectedEvents, Action action, bool assertExpected, string[] expectedPaths)
+        {
+            bool result;
+
+            string[] expectedFullPaths = expectedPaths == null ? null : expectedPaths.Select(e => Path.GetFullPath(e)).ToArray();
+            AutoResetEvent changed = WatchChanged(watcher, (expectedEvents & WatcherChangeTypes.Changed) > 0 ? expectedPaths : null);
+            AutoResetEvent created = WatchCreated(watcher, (expectedEvents & WatcherChangeTypes.Created) > 0 ? expectedPaths : null);
+            AutoResetEvent deleted = WatchDeleted(watcher, (expectedEvents & WatcherChangeTypes.Deleted) > 0 ? expectedPaths : null);
+            AutoResetEvent renamed = WatchRenamed(watcher, (expectedEvents & WatcherChangeTypes.Renamed) > 0 ? expectedPaths : null);
+
+            watcher.EnableRaisingEvents = true;
+            action();
+
+            // Verify Changed
+            bool Changed_expected = ((expectedEvents & WatcherChangeTypes.Changed) > 0);
+            bool Changed_actual = changed.WaitOne(WaitForExpectedEventTimeout);
```

Comment:
Yes to both questions. We wait for all events to ensure they are thrown (or not thrown) as expected. To make up for this, event waits after the first Changed are done using a shorter timeout value (10 ms vs 50 ms for Changed).

All of the timeout values are much lower than they were before this PR (30000/500 ms for expected/unexpected before) so there's still a net decrease in execution time. Adding the retry functionality lets us use a lower timeout without worrying as much about failures from CI machine slowdown, so the average execution is going to be faster.

I wouldn't say they're unnecessary delays though; we should be verifying that a test isn't throwing an unexpected event type.